File: Kernel Simulator/Misc/TextEdit/Commands/QueryWordRegex.vb

'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports Extensification.IntegerExts
Imports KS.Misc.Reflection

Namespace Misc.TextEdit.Commands
    Class TextEdit_QueryWordRegexCommand
        Inherits CommandExecutor
        Implements ICommand

        Public Overrides Sub Execute(StringArgs As String, ListArgs() As String, ListArgsOnly As String(), ListSwitchesOnly As String()) Implements ICommand.Execute
            If ListArgs?.Count = 2 Then
                If IsStringNumeric(ListArgs(1)) Then
                    If CInt(ListArgs(1)) <= TextEdit_FileLines.Count Then
                        Dim QueriedChars As Dictionary(Of Integer, String) = TextEdit_QueryWordRegex(ListArgs(0), ListArgs(1))
                        For Each WordIndex As Integer In QueriedChars.Keys
                            TextWriterColor.Write("- {0}: ", False, ColTypes.ListEntry, WordIndex)
                            TextWriterColor.Write("{0} ({1})", True, ColTypes.ListValue, ListArgs(0), TextEdit_FileLines(ListArgs(1)))
                        Next
                    Else
                        TextWriterColor.Write(DoTranslation("The specified line number may not be larger than the last file line number."), True, ColTypes.Error)
                    End If
                ElseIf ListArgs(1).ToLower = "all" Then
                    Dim QueriedWords As Dictionary(Of Integer, Dictionary(Of Integer, String)) = TextEdit_QueryWordRegex(ListArgs(0))
                    For Each LineIndex As Integer In QueriedWords.Keys
                        For Each WordIndex As Integer In QueriedWords(LineIndex).Keys
                            TextWriterColor.Write("- {0}:{1}: ", False, ColTypes.ListEntry, LineIndex, WordIndex)
                            TextWriterColor.Write("{0} ({1})", True, ColTypes.ListValue, ListArgs(0), TextEdit_FileLines(LineIndex))
                        Next
                    Next
                End If
            ElseIf ListArgs?.Count > 2 Then
                If IsStringNumeric(ListArgs(1)) And IsStringNumeric(ListArgs(2)) Then
                    If CInt(ListArgs(1)) <= TextEdit_FileLines.Count And CInt(ListArgs(2)) <= TextEdit_FileLines.Count Then
                        Dim LineNumberStart As Integer = ListArgs(1)
                        Dim LineNumberEnd As Integer = ListArgs(2)
                        LineNumberStart.SwapIfSourceLarger(LineNumberEnd)
                        For LineNumber = LineNumberStart To LineNumberEnd
                            Dim QueriedChars As Dictionary(Of Integer, String) = TextEdit_QueryWordRegex(ListArgs(0), LineNumber)
                            For Each WordIndex As Integer In QueriedChars.Keys
                                TextWriterColor.Write("- {0}:{1}: ", False, ColTypes.ListEntry, LineNumber, WordIndex)
                                TextWriterColor.Write("{0} ({1})", True, ColTypes.ListValue, ListArgs(0), TextEdit_FileLines(ListArgs(1)))
                            Next
                        Next
                    Else
                        TextWriterColor.Write(DoTranslation("The specified line number may not be larger than the last file line number."), True, ColTypes.Error)
                    End If
                End If
            End If
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Editors/TextEdit/Commands/QueryWordRegex.cs
using System;
using System.Linq;
using KS.ConsoleBase.Colors;
using KS.Languages;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using KS.Misc.Reflection;
using KS.Misc.Text;
using KS.Misc.Writers.ConsoleWriters;
using KS.Shell.ShellBase.Commands;

namespace KS.Misc.Editors.TextEdit.Commands
{
	class TextEdit_QueryWordRegexCommand : CommandExecutor, ICommand
	{

		public override void Execute(string StringArgs, string[] ListArgs, string[] ListArgsOnly, string[] ListSwitchesOnly)
		{
			if ((ListArgs?.Count()) is { } arg1 && arg1 == 2)
			{
				if (StringQuery.IsStringNumeric(ListArgs[1]))
				{
					if (Convert.ToInt32(ListArgs[1]) <= TextEditShellCommon.TextEdit_FileLines.Count)
					{
						var QueriedChars = TextEditTools.TextEdit_QueryWordRegex(ListArgs[0], Convert.ToInt32(ListArgs[1]));
						foreach (int WordIndex in QueriedChars.Keys)
						{
							TextWriterColor.Write("- {0}: ", false, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListEntry), WordIndex);
							TextWriterColor.Write("{0} ({1})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListValue), ListArgs[0], TextEditShellCommon.TextEdit_FileLines[Convert.ToInt32(ListArgs[1])]);
						}
					}
					else
					{
						TextWriterColor.Write(Translate.DoTranslation("The specified line number may not be larger than the last file line number."), true, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Error));
					}
				}
				else if (ListArgs[1].ToLower() == "all")
				{
					var QueriedWords = TextEditTools.TextEdit_QueryWordRegex(ListArgs[0]);
					foreach (int LineIndex in QueriedWords.Keys)
					{
						foreach (int WordIndex in QueriedWords[LineIndex].Keys)
						{
							TextWriterColor.Write("- {0}:{1}: ", false, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListEntry), LineIndex, WordIndex);
							TextWriterColor.Write("{0} ({1})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListValue), ListArgs[0], TextEditShellCommon.TextEdit_FileLines[LineIndex]);
						}
					}
				}
			}
			else if ((ListArgs?.Count()) is { } arg2 && arg2 > 2)
			{
				if (StringQuery.IsStringNumeric(ListArgs[1]) & StringQuery.IsStringNumeric(ListArgs[2]))
				{
					if (Convert.ToInt32(ListArgs[1]) <= TextEditShellCommon.TextEdit_FileLines.Count & Convert.ToInt32(ListArgs[2]) <= TextEditShellCommon.TextEdit_FileLines.Count)
					{
						int LineNumberStart = Convert.ToInt32(ListArgs[1]);
						int LineNumberEnd = Convert.ToInt32(ListArgs[2]);
						LineNumberStart.SwapIfSourceLarger(ref LineNumberEnd);
						for (int LineNumber = LineNumberStart, loopTo = LineNumberEnd; LineNumber <= loopTo; LineNumber++)
						{
							var QueriedChars = TextEditTools.TextEdit_QueryWordRegex(ListArgs[0], LineNumber);
							foreach (int WordIndex in QueriedChars.Keys)
							{
								TextWriterColor.Write("- {0}:{1}: ", false, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListEntry), LineNumber, WordIndex);
								TextWriterColor.Write("{0} ({1})", true, color: KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.ListValue), ListArgs[0], TextEditShellCommon.TextEdit_FileLines[Convert.ToInt32(ListArgs[1])]);
							}
						}
					}
					else
					{
						TextWriterColor.Write(Translate.DoTranslation("The specified line number may not be larger than the last file line number."), true, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Error));
					}
				}
			}
		}

	}
}
